pkg/gardenlet/controller/managedseed/managedseed_reconciler.go

```diff
@@ -0,0 +1,230 @@
+	"github.com/gardener/gardener/pkg/logger"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/retry"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+// reconciler implements the reconcile.Reconciler interface for ManagedSeed reconciliation.
+type reconciler struct {
+	gardenClient kubernetes.Interface
+	actuator     Actuator
+	recorder     record.EventRecorder
+	logger       *logrus.Logger
+}
+
+// newReconciler creates a new ManagedSeed reconciler with the given clients, actuator, recorder, and logger.
+func newReconciler(gardenClient kubernetes.Interface, actuator Actuator, recorder record.EventRecorder, logger *logrus.Logger) reconcile.Reconciler {
+	return &reconciler{
+		gardenClient: gardenClient,
+		actuator:     actuator,
+		recorder:     recorder,
+		logger:       logger,
+	}
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	managedSeed := &seedmanagementv1alpha1.ManagedSeed{}
+	if err := r.gardenClient.DirectClient().Get(ctx, request.NamespacedName, managedSeed); err != nil {
+		if apierrors.IsNotFound(err) {
+			r.logger.Debugf("Skipping ManagedSeed %s because it has been deleted", request.NamespacedName)
+			return reconcile.Result{}, nil
+		}
+		r.logger.Errorf("Could not get ManagedSeed %s from store: %+v", request.NamespacedName, err)
+		return reconcile.Result{}, err
+	}
+
+	managedSeedLogger := logger.NewFieldLogger(r.logger, "managedSeed", kutil.ObjectName(managedSeed))
+
+	if managedSeed.DeletionTimestamp != nil {
+		return r.delete(ctx, managedSeed, managedSeedLogger)
+	}
+	return r.reconcile(ctx, managedSeed, managedSeedLogger)
+}
+
+func (r *reconciler) reconcile(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, logger *logrus.Entry) (reconcile.Result, error) {
+
+	// Ensure gardener finalizer
+	if err := controllerutils.PatchFinalizers(ctx, r.gardenClient.Client(), managedSeed, gardencorev1beta1.GardenerName); err != nil {
+		logger.Errorf("Could not ensure gardener finalizer: %+v", err)
+		return reconcile.Result{}, err
+	}
+
+	conditionShootExists := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedShootExists)
+	conditionShootReconciled := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedShootReconciled)
+	conditionSeedRegistered := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedSeedRegistered)
+	conditionSeedBootstrapped := gardencorev1beta1helper.GetOrInitCondition(managedSeed.Status.Conditions, seedmanagementv1alpha1.ManagedSeedSeedBootstrapped)
```

Comment:
Agreed in principle, but I think this is very minor and I wouldn't complicate the logic further just because of this. What's actually wrong with the above? It clearly states an "Unknown" status. Also, the current behavior is consistent with other Gardener resources.

Reply:
We agreed it would be nicer to have `False` instead of `Unknown`, with a better message and reason. This should be easy to do.

Reply:
After taking a closer look, while this it is true for some conditions that we can be sure they are `False` if they are missing in the status (e.g. for `SeedRegistered`), it's not true for others (e.g. for `ShootExists` or `ShootReconciled`). I would like to avoid handling different conditions in different ways though.

Therefore, let's do what was initially proposed: unless a condition has been initialized with something different from `Unknown`, it shouldn't be added to the status at all.

Reply:
Done